### PR TITLE
feat: Add an option to skip entries

### PR DESCRIPTION
**Pull request by jan-auer**
This implements a feature request from #37 that allows to skip certain entry points via the `entry` option. The actual feature commit is 67901a7.

Since the code was not yet documented and test coverage was pretty low, I also refactored the whole thing and added more tests. It **should** now better live up to our usual coding standards. Test coverage is at nearly 100%, except for the dry-run CLI mock. 

Tests have been added for our modifications to the compiler options (namely `entry` and `module.rules`/`module.loaders`), which were previously untested. This should also make it easier to test certain webpack configurations, once we receive them from users.

### CODE REVIEW

**Comment by jan-auer** [comment]
@kamilogorek any idea what's going on with codecov there?

File: src/index.js [hunk 1]
```diff
@@ -1,135 +1,216 @@
...
+    return (this.options.release
+      ? Promise.resolve(this.options.release)
+      : this.cli.releases.proposeVersion()
+    ).then(version => `${version}`.trim());
+  }
+
+  /** Checks if the given named entry point should be handled. */
+  checkEntry(key) {
+    const { entries } = this.options;
+    if (!entries) {
```
Comment (kamilogorek) [inline src/index.js]:
```js
{
  entries: '',
  entries: 0
}
```

^ those won't throw an error and pass all entries

File: src/index.js [hunk 2]
```diff
@@ -1,135 +1,216 @@
...
-  if (typeof compiler.options === 'undefined') {
-    // Gatekeeper because we are not running in webpack env
-    // probably just tests
-    return;
+    if (entries instanceof RegExp) {
+      return entries.test(key);
+    }
+
+    if (Array.isArray(entries)) {
+      return entries.indexOf(key) >= 0;
```
Comment (kamilogorek) [inline src/index.js]: We can use `return entries.includes(key);` to be more descriptive

File: src/index.js [hunk 3]
```diff
@@ -1,135 +1,216 @@
...
-  if (typeof changedCompiler.options.module === 'undefined') {
-    changedCompiler.options.module = {};
+
+  /** Injects the release string into the given entry point. */
+  injectEntry(originalEntry, newEntry) {
+    if (Array.isArray(originalEntry)) {
+      return [newEntry].concat(originalEntry);
+    }
+
+    if (originalEntry !== null && typeof originalEntry === 'object') {
```
Comment (kamilogorek) [inline src/index.js]: I know it's nitpicking, but `originalEntry = ['someEntry']` will pass this check as well, and I'm pretty sure we just want plain objects

File: src/index.js [hunk 4]
```diff
@@ -1,135 +1,216 @@
...
+      });
+      return nextEntries;
+    }
+
+    if (typeof originalEntry === 'string') {
+      return [newEntry, originalEntry];
+    }
+
+    if (typeof originalEntry === 'function') {
+      return () =>
```
Comment (kamilogorek) [inline src/index.js]: We can just return `Promise.resolve`, without wrapping it in an intermediate function

File: src/__tests__/index.spec.js [hunk 5]
```diff
@@ -67,156 +77,348 @@ describe('new SentryCliPlugin(options)', () => {
   });
 });
 
-describe('.apply', () => {
+describe('CLI configuration', () => {
+  test('uses a mock SentryCLI in `dryRun` mode', () => {
```
[the remark this answers is not in the document]
Comment (jan-auer) [reply src/__tests__/index.spec.js]: I'll rename the test to make it more clear

File: src/index.js (hunk 2 above)
Comment (jan-auer) [reply src/index.js]: Ah right, for some reason I wanted to be Node anno 1980 compatible -.-

File: src/index.js (hunk 3 above)
Comment (jan-auer) [reply src/index.js]: Can you clarify please?

File: src/index.js (hunk 4 above)
Comment (jan-auer) [reply src/index.js]: Which means that we also have to support Promises in addition to the other kinds of entry records, yes?

File: src/index.js (hunk 3 above)
Comment (kamilogorek) [reply src/index.js]: If `originalEntry` is an array (I'm not even sure if it's possible at this point of the code, so you can just ignore me if it's not), then it'll still execute this code, as `(originalEntry !== null && typeof originalEntry === 'object')` check doesn't prevent it from doing so.

```js
var originalEntry = ['foo'];
console.log(originalEntry !== null && typeof originalEntry === 'object'); // true
```

Which mean that you'll do:

```js
Object.keys(originalEntry).forEach(key => {
  ...
});
```

for every single index, as `Object.keys(array)` returns an array of available indexes, eg. `Object.keys(['a', 'b', 'c']) => [0, 1, 2]`

File: src/index.js (hunk 4 above)
Comment (kamilogorek) [reply src/index.js]: Ah, this was the reason. Ignore me then.

File: src/index.js (hunk 3 above)
Comment (jan-auer) [reply src/index.js]: Ah okay. So if originalEntry is an array, it will be handled by the branch above already. Thanks for the update!

Comment (kamilogorek) [reply src/index.js]: Right... 🤦‍♂️

**Pull request merged**